### PR TITLE
[SVCS-448] Add additional tests and fixtures to S3.

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Our S3 test coverage is pretty good, but not perfect. This PR adds tests for intra_copies and elaborates on many edge cases for S3 metadata returns.

## Changes

- Coverage for intra_move/copy
- Coverage for confirm_delete
- Coverage for deleting empty folders 
- Coverage getting metadata folders with single items
- Coverage getting metadata revisions with a single version

## Side Effects 

Some existing tests were split up for clarity.

## Ticket

https://openscience.atlassian.net/browse/SVCS-448

### CODE REVIEW

File: tests/providers/s3/test_provider.py [hunk 1]
```diff
@@ -2,10 +2,13 @@
 
 from tests.utils import MockCoroutine
 
+
+import os
 import io
 import time
 import base64
 import hashlib
+from urllib import parse
```
Comment (felliott) [inline tests/providers/s3/test_provider.py]: Nitpick: order imports by line length

File: tests/providers/s3/test_provider.py [hunk 2]
```diff
@@ -716,7 +884,7 @@ class TestMetadata:
 
     @pytest.mark.asyncio
     @pytest.mark.aiohttpretty
-    async def test_upload(self, provider, file_content, file_stream, file_metadata, mock_time):
+    async def test_upload(self, provider, file_content, file_stream, file_header_metadata):
```
Comment (felliott) [inline tests/providers/s3/test_provider.py]: Why did you take out `mock_time` here (and in a bunch of other places)?  Is it not needed anymore?

File: tests/providers/s3/test_metadata.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        assert file_metadata_object.modified == '2009-10-12T17:50:30.000Z'
+        assert file_metadata_object.etag == "fba9dede5f27731c9771645a39863328".replace('"', '')
+        assert not file_metadata_object.created_utc
+        assert not file_metadata_object.content_type
+
+class TestFolderKeyMetadata:
+
+    def test_folder_key_metadata(self, folder_key_metadata_object):
+        folder_key_metadata_object.name == 'naptime'
+        folder_key_metadata_object.path == '/naptime'
```
Comment (felliott) [inline tests/providers/s3/test_metadata.py]: Missing `assert`s!

File: tests/providers/s3/test_provider.py (hunk 2 above)
Comment (Johnetordoff) [reply tests/providers/s3/test_provider.py]: Doesn't appear to be necessary in all cases.

File: tests/providers/s3/test_metadata.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        assert file_metadata_object.name == 'my-image.jpg'
+        assert file_metadata_object.path == '/my-image.jpg'
+        assert file_metadata_object.materialized_path == '/my-image.jpg'
+        assert file_metadata_object.size == 434234
+        assert file_metadata_object.modified == '2009-10-12T17:50:30.000Z'
+        assert file_metadata_object.etag == "fba9dede5f27731c9771645a39863328".replace('"', '')
+        assert not file_metadata_object.created_utc
+        assert not file_metadata_object.content_type
+
+        md5 = file_metadata_object.raw['ETag'].replace('"', '')
```
Comment (felliott) [inline tests/providers/s3/test_metadata.py]: Hardcode this, please.

File: tests/providers/s3/fixtures/fixtures.py [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+    }
+
+
+@pytest.fixture
+def file_content():
+    return b'sleepy'
+
+@pytest.fixture
+def folder_metadata():
+    return b'''<?xml version="1.0" encoding="UTF-8"?>
```
Comment (felliott) [inline tests/providers/s3/fixtures/fixtures.py]: Please move these into `fixtures/foo.xml` files, similar to how we do JSON.

File: tests/providers/s3/fixtures/fixtures.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    content = OrderedDict(Key='naptime/',
+                          LastModified='2009-10-12T17:50:30.000Z',
+                          ETag='"fba9dede5f27731c9771645a39863328"',
+                          Size='0',
+                          StorageClass='STANDARD')
+
+
+    folder = S3FolderKeyMetadata(content)
+    folder
+    return S3FolderKeyMetadata(content)
```
Comment (felliott) [inline tests/providers/s3/fixtures/fixtures.py]: ☹️

File: tests/providers/s3/test_metadata.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+)
+
+class TestFileMetadataHeaders:
+
+    def test_file_metadata_headers(self, file_metadata_headers_object, file_header_metadata):
+        assert not file_metadata_headers_object.is_folder
+        assert file_metadata_headers_object.is_file
+        assert file_metadata_headers_object.name == 'test-path'
+        assert file_metadata_headers_object.path == '//test-path'
+        assert file_metadata_headers_object.materialized_path == '//test-path'
```
Comment (felliott) [inline tests/providers/s3/test_metadata.py]: Red flag: why does this and `path` begin with two slashes?  If this is legit, it's probably a bug.

File: tests/providers/s3/test_metadata.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        assert file_metadata_headers_object.kind == 'file'
+        assert file_metadata_headers_object.provider == 's3'
+        assert file_metadata_headers_object.size == file_header_metadata['CONTENT-LENGTH']
+        assert file_metadata_headers_object.content_type == file_header_metadata['CONTENT-TYPE']
+        assert file_metadata_headers_object.modified == file_header_metadata['LAST-MODIFIED']
```
Comment (felliott) [inline tests/providers/s3/test_metadata.py]: Hardcode these, please.

Comment (felliott) [reply tests/providers/s3/test_metadata.py]: ...and the others.

File: tests/providers/s3/test_metadata.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        md5 = file_metadata_headers_object.raw['ETAG'].replace('"', '')
+        extra = {
+            'md5': md5,
+            'encryption': file_metadata_headers_object.raw.get('X-AMZ-SERVER-SIDE-ENCRYPTION', ''),
+            'hashes': {
+                'md5': md5,
+            },
+        }
+
+        assert file_metadata_headers_object.extra == extra
```
Comment (felliott) [inline tests/providers/s3/test_metadata.py]: Please add a test for `json_api_serialized` a la https://github.com/CenterForOpenScience/waterbutler/blob/develop/tests/providers/box/test_metadata.py#L55

File: tests/providers/s3/test_provider.py [hunk 10]
```diff
@@ -651,7 +566,7 @@ class TestMetadata:
         assert result[0].name == '   photos'
         assert result[1].name == 'my-image.jpg'
         assert result[2].extra['md5'] == '1b2cf535f27731c974343645a3985328'
-        assert result[2].extra['hashes']['md5'] == '1b2cf535f27731c974343645a3985328'
+        assert result[2].extra['hashes']['md5'] == '1b2cf535f27731c974343645a3985328'######################
```
Comment (felliott) [inline tests/providers/s3/test_provider.py]: ???

File: tests/providers/s3/test_provider.py (hunk 2 above)
Comment (felliott) [reply tests/providers/s3/test_provider.py]: It fixes a probabilistic test failure in the signature calculator.  If you calculate it to close to a clock tick, your generated and expected signatures won't match.  I got a test failure my first time running this, though you might have to do it a few times to see it.

File: tests/providers/s3/fixtures/fixtures.py (hunk 5 above)
Comment (Johnetordoff) [reply tests/providers/s3/fixtures/fixtures.py]: Should I create an independent xml file for each fixture? If I put it all in one file, then parse as XML it like the JSON I'll then have to convert it back into bytes to use it.

Comment (felliott) [reply tests/providers/s3/fixtures/fixtures.py]: Yep! Independent files, please.

File: tests/providers/s3/test_metadata.py (hunk 7 above)
Comment (Johnetordoff) [reply tests/providers/s3/test_metadata.py]: Yep, looks like a bug, should I get it in this PR or next ticket?

Comment (felliott) [reply tests/providers/s3/test_metadata.py]: New ticket, please.
